CesiumGltf/include/CesiumGltf/PropertyArrayView.h

```diff
@@ -34,49 +36,91 @@ template <typename ElementType> class PropertyArrayView {
   PropertyArrayView(const gsl::span<const std::byte>& buffer) noexcept
       : _values{CesiumUtility::reintepretCastSpan<const ElementType>(buffer)} {}
 
+  const ElementType& operator[](int64_t index) const noexcept {
+    return this->_values[index];
+  }
+
+  int64_t size() const noexcept { return this->_values.size(); }
+
+  auto begin() { return this->_values.begin(); }
+  auto end() { return this->_values.end(); }
+  auto begin() const { return this->_values.begin(); }
+  auto end() const { return this->_values.end(); }
+
+private:
+  gsl::span<const ElementType> _values;
+};
+
+/**
+ * @brief A view on an array element of a {@link PropertyTableProperty}
+ * or {@link PropertyTextureProperty}.
+ *
+ * Provides utility to retrieve the data stored in the array of
+ * elements via the array index operator.
+ */
+template <typename ElementType> class PropertyArrayCopy {
```

Comment:
This is definitely helpful, but I'm curious what scenario it was solving (just so I know where I broke things :sweat_smile:)

Reply:
Nah you didn't do anything wrong! The reason for this comes down to this note inn the [swl-variant](https://github.com/groundswellaudio/swl-variant) readme:

> Clang 12/13 will only work for trivially destructible types as their implementation of C++20 is incomplete.

We're not using `swl::variant` anywhere in cesium-native (yet), but we're using it in both Unreal and Unity in the vcpkg branches in order to drastically improve build times. But anywhere we replace `std::variant` with `swl::variant`, we need to watch out for this limitation that all of the types in the variant must be trivially destructible (roughly: they don't have an explicit destructor, and neither do any of their fields).

`std::vector` isn't trivially destructible because it needs to free the heap-allocated memory it uses to hold the elements, and so neither is anything that includes `std::vector`. The old implementation of `PropertyArrayView` wasn't trivially destructible either because it held a vector. Now, PropertyArrayView is trivially destructible, and that's what we use in the `swl::variant`. PropertyArrayCopy exists for the cases where we need ownership semantics.

Then we do dirty tricks, like this, from CesiumGS/cesium-unreal#1447:

```cpp
  template <typename ArrayType>
  explicit FCesiumMetadataValue(CesiumGltf::PropertyArrayCopy<ArrayType>&& Copy)
      : _value(), _valueType(), _storage() {
    this->_value = std::move(Copy).toViewAndExternalBuffer(this->_storage);

    ECesiumMetadataType type =
        ECesiumMetadataType(CesiumGltf::TypeToPropertyType<ArrayType>::value);
    ECesiumMetadataComponentType componentType = ECesiumMetadataComponentType(
        CesiumGltf::TypeToPropertyType<ArrayType>::component);
    bool isArray = true;
    this->_valueType = {type, componentType, isArray};
  }
```

When we construct a metadata value from a _copy_ we convert it to a _view_ plus external storage, and store each separately. That way metadata value's variant can be a swl::variant`.

That's a lot of words and still not very clear, so let me know if it makes any sense.

Reply:
I think I get it! I did notice the "hot potato" you were playing with the memory, but it all makes sense now 😛